src/core/lib/deck.js

```diff
@@ -160,13 +230,51 @@ export default class Deck {
     return canvas;
   }
 
+  _createControls(props) {
+    this.eventManager = new EventManager(this.canvas);
+
+    // If props.controls is not provided, fallback to default MapControls instance
+    // Cannot use defaultProps here because it needs to be per map instance
+    let controls = props.controls;
+    if (props.viewportState) {
+      controls = new ViewportControls(props.viewportState, props);
+    }
+    if (controls) {
+      controls.setOptions(
+        Object.assign({}, props, props.viewState, {
+          onStateChange: this._onInteractiveStateChange.bind(this),
+          eventManager: this.eventManager,
+          // Add a default event handler in case non was provided
+          // This allows basic examples to not specify any callbacks
+          onViewportChange: props.onViewportChange || this._onViewportChange.bind(this)
```

Comment:
I do not like how the purpose of this prop is overloaded. Consider if an application needs to render some controls synchronized with deck.gl. The name of `onViewportChange` suggests that it behaves as an event listener. Subscribing to an event should not change how a component behaves.

At the least, the callback should offer the user a way to retain the default behavior. E.g.
```
onViewportChange={
  viewport => {
    // do something
    const newViewport = {
      ...viewport,
      longitude: clamp(viewport.longitude, -100, -80)
    };
    return newViewport;  // continue internal state update
  }
}
```
This way we are still backwards compatible (existing apps do not return a value at the end of the handler) and can support auto state update by assigning the default prop `onViewportChange: viewport=>viewport`.